rust/datafusion/src/optimizer/utils.rs

```diff
@@ -0,0 +1,184 @@
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Collection of utility functions that are leveraged by the query optimizer rules
+
+use arrow::datatypes::{DataType, Field, Schema};
+
+use crate::error::{ExecutionError, Result};
+use crate::logicalplan::Expr;
+
+/// Create field meta-data from an expression, for use in a result set schema
+pub fn expr_to_field(e: &Expr, input_schema: &Schema) -> Result<Field> {
+    match e {
+        Expr::Column(i) => Ok(input_schema.fields()[*i].clone()),
+        Expr::Literal(ref lit) => Ok(Field::new("lit", lit.get_datatype()?, true)),
+        Expr::ScalarFunction {
+            ref name,
+            ref return_type,
+            ..
+        } => Ok(Field::new(&name, return_type.clone(), true)),
+        Expr::AggregateFunction {
+            ref name,
+            ref return_type,
+            ..
+        } => Ok(Field::new(&name, return_type.clone(), true)),
+        Expr::Cast { ref data_type, .. } => {
+            Ok(Field::new("cast", data_type.clone(), true))
+        }
+        Expr::BinaryExpr {
+            ref left,
+            ref right,
+            ..
+        } => {
+            let left_type = left.get_type(input_schema)?;
+            let right_type = right.get_type(input_schema)?;
+            Ok(Field::new(
+                "binary_expr",
+                get_supertype(&left_type, &right_type).unwrap(),
+                true,
+            ))
+        }
+        _ => Err(ExecutionError::NotImplemented(format!(
+            "Cannot determine schema type for expression {:?}",
+            e
+        ))),
+    }
+}
+
+/// Create field meta-data from an expression, for use in a result set schema
+pub fn exprlist_to_fields(expr: &Vec<Expr>, input_schema: &Schema) -> Result<Vec<Field>> {
+    expr.iter()
+        .map(|e| expr_to_field(e, input_schema))
+        .collect()
+}
+
+/// Given two datatypes, determine the supertype that both types can safely be cast to
+pub fn get_supertype(l: &DataType, r: &DataType) -> Result<DataType> {
+    match _get_supertype(l, r) {
+        Some(dt) => Ok(dt),
+        None => match _get_supertype(r, l) {
+            Some(dt) => Ok(dt),
+            None => Err(ExecutionError::InternalError(format!(
+                "Failed to determine supertype of {:?} and {:?}",
+                l, r
+            ))),
+        },
+    }
+}
+
+/// Given two datatypes, determine the supertype that both types can safely be cast to
+fn _get_supertype(l: &DataType, r: &DataType) -> Option<DataType> {
```

Comment:
This is pretty error prone. Something like [this](https://github.com/ibis-project/ibis/blob/master/ibis/expr/datatypes.py#L1445-L1477) would be better.

Reply:
This should't block this PR, but We should have a ticket about it.

Reply:
I agree. That is much nicer.

Reply:
Can we put all these types in an array or other kind of data structure and do lookups to determine the least-common-ancestor?

Reply:
Filed https://issues.apache.org/jira/browse/ARROW-4957 to track this